### DISCUSSION TITLE
ESP8266 Getting disconnected from WiFi Automatically. SDK issue.

### DISCUSSION THREAD

**Discussion by Asanga-Viraj** [SmingHub/Sming#2510] [Q&A] [answered]
Hello Sming Team,

I experienced a issue with ESP8266. It disconnects from the WiFi network and it doesn't connect automatically after number of days. I did some search on internet and found that it as a memory leak on SDK (with time of days) and ESP doesn't identify WiFi connectivity (ES8266 still thinks WiFi is connected). The suggestions is check WifiStation.isConnected() and do restart ESP if value is false. I am not sure about this because during this issue ESP thinks still Station is connected. 

I am thinking about more robust system. Ping the router frequently and restart the ESP if it failed. I found Sming inbuilt resource ping.h on "Sming\Arch\Esp8266\Components\esp-open-lwip\esp-open-lwip\lwip\app" directory. But they are not available to application level (ping.o is not available). Can you help me with this.

Do you have any other solution for this?

Thank you in advance.

Regards,
Asanga

**Comment by slaff** [comment]
> I did some search on internet and found that it as a memory leak on SDK (with time of days) 

Can you please paste the link to this information here? The best would be if we can fix this memory leak.

> I am thinking about more robust system. Ping the router frequently and restart the ESP if it failed. 

It might work but usually memory leaks tend to eat most of the available memory on an embedded device. So it is not clear if the device will have enough memory to execute even ping...

> But they are not available to application level (ping.o is not available).

ICPM packets are allowed for ESP8266  in both LWIP versions that are available.

  https://github.com/pfalcon/esp-open-lwip/blob/b9a23a6e9ef88d995ae74014e71a6a4227f61d62/include/lwipopts.h#L596-L601

So what is missing is a sample implementation. I came upon [this implementation](https://github.com/dreamcat4/lwip/blob/master/contrib/apps/ping/ping.c) which should compile and work also on Sming.

Do you want to try it and add it as an improvement to Sming?

**Reply by Asanga-Viraj** [reply]
Thank you very much slaff.

I totally agree. I couldn't say much information about this because it is totally random and it takes days to happen. This might not be a memory leak but it breaks WiFi station mode. I added the following links as references.

https://arduino.stackexchange.com/questions/73182/nodemcu-esp8266-disconnected-from-wifi-after-several-hours-running
https://community.mongoose-os.com/t/esp8266-disconnects-from-wifi-and-doesnt-reconnect-automatically/2026
https://github.com/esp8266/Arduino/issues/5083
https://github.com/platformio/platform-espressif8266/issues/135

Yes, I want to try it. Not enough memory to do a ping means something wrong with ESP. So, it is better to reset it. It would be better if you can add [your implementation](https://github.com/dreamcat4/lwip/blob/master/contrib/apps/ping/ping.c) as an improvement to Sming.

**Comment by slaff** [comment, marked as answer]
@Asanga-Viraj You can try PR #2513 which contains network ping sample for ESP8266.

**Reply by Asanga-Viraj** [reply]
Thank you very much @slaff . This is my exact answer.
